staging/src/kubevirt.io/api/core/v1/schema.go

```diff
@@ -465,6 +465,13 @@ type Devices struct {
 	// Whether to emulate a TPM device.
 	// +optional
 	TPM *TPMDevice `json:"tpm,omitempty"`
+
+	// The number of PCIe ports available for the VM. A value lower than the
+	// required amount will get bumped by libvirt to same as amount of inserted
+	// PCIe cards.
+	// Defaults to 0.
+	// +optional
+	NumberPciPorts uint8 `json:"numberPciPorts,omitempty"`
```

Comment:
I find the introduction of this problematic for several reasons:
- The suggested API attempts to abstract the definition of controllers of PCI devices. But seems to be dependent on many factors, like the machine-type and possibly other controllers that get defined by default. If we expose controllers, we may need to go much deeper in the details (or allow details to be added later).
- Libvirt documentation warns about odd behavior which may not provided the expected results:
> Note that if you're adding PCI controllers to a guest and at the same time you're also adding PCI devices, some of the controllers will be used for the newly-added devices and won't be available for hotplug once the guest has been started.

I would suggest extracting this to a separate PR and bring into the discussion the compute SIG and libvirt experts.

Reply:
This is what is done for both rhv and openstack, and without this, users get to hotplug a single interface in. 

That's imo pretty useless. 

Furthermore, I don't see how what you pasted applies: users do not get to update this value - this is only applied when the domain is created.

Reply:
Users have no visibility to the number of controller they need.
We do not reflect/expose the required controllers given a specific configuration.
Therefore, the API proposed is hard to use for users, hard to extend (e.g. what if we want to control a property of the controller in the future?) and too low level (i.e. taking over the libvirt automatic controller assignment and asking the users to own it).

The feature itself is not production ready, I think we are good with having a limitation at the moment and following this up with alternatives we can discuss at a later stage.
I place to write a suggestion soon.

If we take this change out of this PR, it will also allow a more focus conversation involving other teams.